src/storage_manager.rs

```diff
@@ -56,7 +56,12 @@ pub trait StorageManagerInterface: Send + Sync {
     ///
     /// Callback function pointer to native (kotlin/swift) code for
     /// getting a key.
-    fn get(&self, key: Key) -> Result<Option<Value>, StorageManagerError>;
+    fn get(&self, key: Key) -> Result<Value, StorageManagerError>;
+
+    /// Function: list
+    ///
+    /// Callback function pointer for listing available keys.
+    fn list(&self) -> Vec<Key>;
```

Comment:
```suggestion
    fn list(&self) -> Result<Vec<Key>, StorageManagerError>;
```

Reply:
The possible `StorageManagerError` values are:

- invalid key
- unable to decrypt value
- storage full
- couldn't make key

None of these are applicable to getting a list of keys, so it seemed to me that either we needed to extend the error list to cover the only failure case I can think of (the underlying storage manager finding a key/value pair but being somehow unable to read the key, I guess this would be `StorageManagerErrror::SystemError` or somesuch?), or it made more sense to just return a list regardless.

From an API point of view, if the caller is asking for a list, they almost certainly just want to know what is available that can be used, so returning a list of keys seems more robust to me; otherwise we'd be returning `Err(something)` instead of the list if any of the keys were a problem, or returning a `Vec<Result<Key, Error>>` which seems like it would be a hassle to use.

Reply:
In the swift implementation there seems to be an error being ignored here:
https://github.com/spruceid/mobile-sdk-swift/pull/24/files#diff-c794a03170cd0cf34fb4fa8c6f7c69791900504d65c0c769534c579a52be1106R87-R91

If it doesn't fit any of the existing errors, then a new error variant should be added.

Reply:
I agree that `Vec<Result<Key, Error>>` is not helpful.

Reply:
Indeed, but that is likely to be "directory doesn't exist", which would be the result if no keys had been stored yet (the directory is created lazily). In the "no keys created yet" case, at least, I think we want to return an empty list rather than a storage manager error.

I believe `contentsOfDirectory()` is wrapping `opendir()`, so in theory it can return:

- ENOENT -- isn't there, we should return an empty list
- EMFILE, ENFILE, ENOMEM -- OS resource failure, if we hit this state everything's likely gone pear-shaped and we almost certainly have larger problems
- ENOTDIR -- something took our directory name and created a regular file (or a fifo or something?) instead

The "it's not a directory" case is a corner case; something else would have had to create a regular file (or fifo or something) with the directory path we were expecting to use.  On one hand, that seems like an attempted attack; maybe they've symlinked it to some important file and are hoping we nuke it or do something that lets them read it. On the other hand it could be bad luck or misconfiguration.  Or filesystem corruption, I suppose.

For that case I'd be inclined to fix it inside the storage manager and complain to the logger rather than report an error; all the caller could do with an error is throw up a dialog informing the user that things are broken.

Reply:
The problem is if we don't bubble up the error then we will get unexpected behaviour in the frontend.

For example, this code will be used by the OpenID4VP protocol to display to the user which credentials they can present in response to a request. If we just return an empty array when there is an error then we can no longer do any handling on the frontend and in the protocol for that error.

In general I think it is responsible for libraries to bubble up any and all errors, and leave it up to the application to decide if and how it should handle the error.